### ISSUE TITLE
missing rule: require line break after variable assignment

### ISSUE THREAD

**Issue by remitbri** [eslint/eslint#1622]
Hi. I've tried to review each and every rule of JSHint, JSCS and ESLint to see if I could entirely "get rid of" the first two in favour of the latter one. As suggested by the owner of the @geteslint twitter account, I'm reporting the JSCS rules I've found actually not covered by ESLint. My apologies if it's a duplicate.

- "requireLineBreakAfterVariableAssignment"

E.g this is invalid

```js
var abc = 8; var foo = 5;

var a, b, c,
    foo = 7, bar = 8;
```

**Comment by xjamundx** [comment]
https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/require-line-break-after-variable-assignment.js

**Comment by doberkofler** [comment]
+1

**Comment by lo1tuma** [comment]
Should this only apply to assignments within a variable declaration?

So this would be valid:
```js
var a; var b; var c;

var a; var b = 42;

var a, b = 42,
    c = 42;
```

But this would be invalid:
```js
var a; var b = 42; var c;

var a, b = 42, c;
```

Shouldn’t there be an option to enfore linebreaks after each delaration no matter if they have a value assigned or not?

BTW: I think the existing rule `newline-after-var` could be easily confused with this rule.

**Comment by nzakas** [comment]
Maybe just make this an option for `newline-after-var`?

**Comment by gyandeeps** [comment]
My 2 cents here:
I think this should be an option in `newline-after-var` rule. 
I think this rule should apply to variable declarations no matter if they have a value assigned or not.

**Comment by mathieumg** [comment]
I think this should follow the JSCS rule (as per the original post), i.e. if there is an assignation it must be alone on the line, otherwise multiple declarations on the same line are fine. I'm speaking regarding a single `var` statement, I think there should never be more than one per line in any case.

Valid:

```js
var a, b, c,
      d = 42;
```

```js
var a = 42,
      b, c, d;
```

Invalid:

```js
var a, b, c = 41,
      d = 42;
```

```js
var a = 42, b = 40, c,
      d = 41;
```

**Comment by gyandeeps** [comment]
But we need to some how support functionality to disallow multiple `var` on same line to. Someone (+ me) is going to as for it in the future.

invalid

```js
var a = 1; var b = 2;
```

valid

```js
var a = 1; 
var b = 2;
```

**Comment by robcolburn** [comment]
+1 @gyandeeps

~~My~~  @cowboy's two cents:
http://benalman.com/news/2012/05/multiple-var-statements-javascript/

**Comment by gyandeeps** [comment]
In my plugin I have created a rule which checks for vars on new line without any exceptions to it.
https://github.com/gyandeeps/eslint-plugin-ideal

**Comment by alberto** [comment]
I think this should be a different rule, since it deals with linebreaks between declarators (or assignments), while `newline-after-var` deals with padding after declaration. Ideally `newline-after-var` should be renamed to `padding-after-var`

**Comment by nzakas** [comment]
You're probably right about a separate rule. Do you have a name suggestion?

**Comment by alberto** [comment]
If we just want to enforce it for assignments, I would say `linebreak-after-assignment` or `newline-after-assignment`

If we want to optionally also enforce it for declarators, like

```
var a,
    b;
```

I'm thinking of `linebreak-after-declaration` or `newline-after-declaration`, with an options `all` and `assignments` (which would behave like the one above).

**Comment by nzakas** [comment]
I think we've hit our max of breaking changes for 2.0.0, so we need to keep the current rule. 

I think `newline-after-assignment` doesn't indicate this is just for variable declaration initialization, so maybe `newline-after-var-init`?

**Comment by alberto** [comment]
Wouldn't this rule then include declarations without initialization, like in my last comment?

**Comment by nzakas** [comment]
If there were not initializations, then the rule would do nothing (unless I'm misunderstanding the request).

**Comment by alberto** [comment]
What I was thinking of this two options (names open for bikeshedding):

## "all"

This would be valid:
```
var a, 
    b,
    c,
    d = 42;

var a = 42,
    b;
```
This would be invalid:
```
var a, b;
var a, b = 40;
var a, b,
    d = 42;
```

## "assignments"

This would be valid:
```
var a, b, c, 
    d = 42;

var a, 
   b = 40,
   c = 41;
```
This would be invalid:
```
var a, b, c = 41;

var a, b, 
    c = 41, d = 42;
```

**Comment by nzakas** [comment]
I like it! Id probably call it "initializations" instead of "assignments", but otherwise, :+1:

**Comment by alberto** [comment]
Just to be sure:

Should the part regarding the validness of multiple var statements be left to `newline-after-var`? (Or even a new rule about multiple statements per line)

```
var a; var b; var c;
```

**Comment by nzakas** [comment]
Yes, leave that to `newline-after-var`.

**Comment by alberto** [comment]
Since newlines must come not only after "initializations" but also before, I think we need a better name.
Some suggestions:
```
newline-per-var-declaration
no-multi-var-declaration-per-line
one-var-declaration-per-line
var-declaration-style
```
Depending on the name, maybe the option names `all` and `initializations` will need adjustment too.

I'm open for suggestions. :)

**Comment by ilyavolodin** [comment]
`var-declaration-style` sounds OK to me.

**Comment by nzakas** [comment]
I'd go for `one- var-declaration-per-line`. The word "style" seems too ambiguous.
